table/tables/partition.go

```diff
@@ -387,157 +558,236 @@ func extractListPartitionExprColumns(ctx sessionctx.Context, pi *model.Partition
 	columns []*expression.Column, names types.NameSlice) (*PartitionExpr, error) {
 	// The caller should assure partition info is not nil.
-	locateExprs := make([]expression.Expression, 0, len(pi.Definitions))
-	pruneExprs := make([]expression.Expression, 0, len(pi.Definitions))
-	schema := expression.NewSchema(columns...)
+	pi := tblInfo.GetPartitionInfo()
 	exprCols, offset, err := extractListPartitionExprColumns(ctx, pi, columns, names)
 	if err != nil {
 		return nil, err
 	}
-	p := parser.New()
-	for _, def := range pi.Definitions {
-		exprStr, err := generateListPartitionExprStr(ctx, pi, schema, names, &def, p)
-		if err != nil {
-			return nil, err
-		}
-		expr, err := parseSimpleExprWithNames(p, ctx, exprStr, schema, names)
-		if err != nil {
-			// If it got an error here, ddl may hang forever, so this error log is important.
-			logutil.BgLogger().Error("wrong table partition expression", zap.String("expression", exprStr), zap.Error(err))
-			return nil, errors.Trace(err)
-		}
-		locateExprs = append(locateExprs, expr)
-		pruneExpr := expr.Clone()
-		cols := expression.ExtractColumns(pruneExpr)
-		for _, c := range cols {
-			idx := findIdxByColUniqueID(exprCols, c)
-			if idx < 0 {
-				panic("should never happen")
-			}
-			c.Index = idx
-		}
-
-		pruneExprs = append(pruneExprs, pruneExpr)
+	listPrune := &ForListPruning{}
+	if len(pi.Columns) == 0 {
+		err = listPrune.buildListPruner(ctx, tblInfo, exprCols, columns, names)
+	} else {
+		err = listPrune.buildListColumnsPruner(ctx, tblInfo, columns, names)
+	}
+	if err != nil {
+		return nil, err
 	}
 	ret := &PartitionExpr{
-		InValues:       locateExprs,
-		ForListPruning: &ForListPruning{Exprs: pruneExprs, ExprCols: exprCols},
+		ForListPruning: listPrune,
 		ColumnOffset:   offset,
 	}
 	return ret, nil
 }
 
-func generateListPartitionExprStr(ctx sessionctx.Context, pi *model.PartitionInfo,
-	schema *expression.Schema, names types.NameSlice, def *model.PartitionDefinition, p *parser.Parser) (string, error) {
-	if len(pi.Columns) < 2 {
-		return generateListColumnsPartitionExprStr(ctx, pi, schema, names, def, p)
+func (lp *ForListPruning) buildListPruner(ctx sessionctx.Context, tblInfo *model.TableInfo, exprCols []*expression.Column,
+	columns []*expression.Column, names types.NameSlice) error {
+	pi := tblInfo.GetPartitionInfo()
+	schema := expression.NewSchema(columns...)
+	p := parser.New()
+	expr, err := parseSimpleExprWithNames(p, ctx, pi.Expr, schema, names)
+	if err != nil {
+		// If it got an error here, ddl may hang forever, so this error log is important.
+		logutil.BgLogger().Error("wrong table partition expression", zap.String("expression", pi.Expr), zap.Error(err))
+		return errors.Trace(err)
 	}
-	return generateMultiListColumnsPartitionExprStr(ctx, pi, schema, names, def, p)
+	// Since need to change the column index of the expresion, clone the expression first.
+	lp.LocateExpr = expr.Clone()
+	lp.PruneExprCols = exprCols
+	lp.PruneExpr = expr.Clone()
```

Comment:
maybe need a join test case like https://github.com/pingcap/tidb/issues/13491

current PR rely on range detacher which rely on matched-unique-id, so we cannot prebuild expr in load table schema stage. (hash prune choose build expr for each prune which is slow,  list prune didn't rely on detacher so avoid this question)

maybe a idea is build a unique-id-sandbox (a env that columid start with 0 to prebuild and call rangedetacher) like https://github.com/pingcap/tidb/pull/20435/files#diff-b186616491c68ad7012d78c7cb5f66f0a0891e18f01892c40e3819d0811d43e2R633..(change range detacher need modify too much code :disappointed: )

Reply:
Great catch. Already add `union select` test case. and there are already some join test case here https://github.com/pingcap/tidb/pull/21577/files#diff-711c993812ab0051c28d65b496b06883b10830005e6496fd4a3b31c5d40498ddR566-R642.

But the list pruner still works. You are right, range detacher which rely on matched-unique-id, so I also change the column unique-id before detacher (here https://github.com/pingcap/tidb/pull/21577/files#diff-4ab18f1fdc8f33ee86effe69e380db16030460da13ee52e6168f56ff281a9e2fR385). The `listPartitionPruner. colIDToUniqueID` help to do this, so it doesn't rely on unique-id now, it rely on the `col.ID`